phoenix-shared/src/commonMain/kotlin/fr.acinq.phoenix/db/cloud/IncomingType.kt

```diff
@@ -0,0 +1,113 @@
+package fr.acinq.phoenix.db.cloud
+
+import fr.acinq.bitcoin.ByteVector32
+import fr.acinq.lightning.db.IncomingPayment
+import fr.acinq.phoenix.db.payments.*
+import kotlinx.serialization.*
+import kotlinx.serialization.cbor.ByteString
+import kotlinx.serialization.cbor.Cbor
+
+@Serializable
+data class IncomingPaymentWrapper @OptIn(ExperimentalSerializationApi::class) constructor(
+    @ByteString
+    val preimage: ByteArray,
+    val origin: OriginWrapper,
+    val received: ReceivedWrapper?,
+    val createdAt: Long
+) {
+    constructor(payment: IncomingPayment): this(
+        preimage = payment.preimage.toByteArray(),
+        origin = OriginWrapper(payment.origin),
+        received = ReceivedWrapper(payment.received),
+        createdAt = payment.createdAt
+    )
+
+    fun unwrap(): IncomingPayment {
+        val unwrappedReceived = received?.let {
+            val originTypeVersion = IncomingOriginTypeVersion.valueOf(origin.type)
+            it.unwrap(originTypeVersion)
+        }
+        return IncomingPayment(
+            preimage = ByteVector32(preimage),
+            origin = origin.unwrap(),
+            received = unwrappedReceived,
+            createdAt = createdAt
+        )
+    }
+
+    @Serializable
+    data class OriginWrapper @OptIn(ExperimentalSerializationApi::class) constructor(
+        val type: String,
+        @ByteString
+        val blob: ByteArray
+    ) {
+        companion object {
+            // constructor
+            operator fun invoke(origin: IncomingPayment.Origin): OriginWrapper {
+                val (type, blob) = origin.mapToDb()
```

Comment:
The blob data generated by the `mapToDb` method is a JSON string, to make it readable and easy to debug, but it's not very efficient as you mentioned, so it would make sense to switch to cbor there as well.

However from the quick test I made using cbor for `IncomingOriginData`, the gain seem marginal, so it may not be worth it.

Reply:
The secret for CBOR is that you have to use **both** `@ByteString` & `ByteArray`:
```
@ByteString
val preimage: ByteArray
```

Otherwise, the space savings is marginal, as you noticed.

Reply:
Even with `@ByteString` the savings seems very small and it would add a lot of code to `IncomingOriginData` (and other similar DB objects).